packages/test/test-end-to-end-tests/src/test/container.spec.ts

```diff
@@ -291,4 +293,104 @@ describeNoCompat("Container", (getTestObjectProvider) => {
+        );
+        assert.strictEqual(
+            container.connectionState, ConnectionState.Connected,
+            "container is not connected when loaded",
+        );
+
+        container.disconnect();
+        assert.strictEqual(container.connectionState, ConnectionState.Disconnected, "container can't disconnect()");
+
+        container.connect();
+        await timeoutPromise(
+            (resolve) => container.once("connected", () => resolve()),
+            {durationMs: timeoutMs, errorMsg: "container connect() timeout"},
+        );
+        assert.strictEqual(container.connectionState, ConnectionState.Connected, "container can't connect()");
+    });
+
+    it("can control op processing with connect() and disconnect()", async () => {
+        const innerRequestHandler = async (request: IRequest, runtime: IContainerRuntimeBase) =>
+                runtime.IFluidHandleContext.resolveHandle(request);
+        const runtimeFactory = (_?: unknown) => new TestContainerRuntimeFactory(
+            TestDataObjectType,
+            getDataStoreFactory(),
+            {},
+            [innerRequestHandler]);
+
+        const localTestObjectProvider = new TestObjectProvider(
+            Loader,
+            provider.driver,
+            runtimeFactory);
+
+        const container1 = await localTestObjectProvider.makeTestContainer() as Container;
+        await timeoutPromise(
+            (resolve) => container1.once("connected", () => resolve()),
+            {durationMs: timeoutMs, errorMsg: "container1 initial connect timeout"},
+        );
+        assert.strictEqual(
+            container1.connectionState, ConnectionState.Connected,
+            "container is not connected when loaded",
+        );
+
+        const dataObject = await requestFluidObject<ITestDataObject>(container1, "default");
+        const directory1 = dataObject._root;
+        directory1.set("key", "value");
+        let value1 = await directory1.get("key");
+        assert.strictEqual(value1, "value", "value1 is not set");
+
+        const container2 = await localTestObjectProvider.loadTestContainer() as Container;
+        await timeoutPromise(
+            (resolve) => container2.once("connected", () => resolve()),
+            {durationMs: timeoutMs, errorMsg: "container2 initial connect timeout"},
+        );
+        const dataObjectTest = await requestFluidObject<ITestDataObject>(container2, "default");
+        const directory2 = dataObjectTest._root;
+        await localTestObjectProvider.ensureSynchronized();
+        let value2 = await directory2.get("key");
+        assert.strictEqual(value2, "value", "value2 is not set");
+
+        container2.disconnect();
+        directory1.set("key", "new-value");
+        value1 = await directory1.get("key");
+        assert.strictEqual(value1, "new-value", "value1 is not changed");
+
+        const valueChangePromise = timeoutPromise(
+            (resolve) => directory2.once("valueChanged", () => resolve()),
```

Comment:
I think you should set up this timeoutPromise before awaiting `directory1.get`.  Suppose disconnect didn't work, then there's a race condition (conceivably, anyway) between the two directories, and maybe directory2's valueChanged event fires before the `directory.get` returns.

Reply:
Actually I'm curious if you can avoid depending on this timeout altogether, since it slows down successful test runs by half a second.  What would `await directory2.get("key")` do?  Any other alternatives?

Reply:
I ran into some issues that led to me to add in the `valueChanged` event check. Basically, if I removed `container2.disconnect()` (simulating disconnect failing), then the test would still pass because there was not enough time for `directory2` to sync with `directory1`. To clarify, see below:

```
directory1.set("key", "new-value");
value1 = await directory1.get("key");  // new-value

value2 = await directory2.get("key");  // value

await waitUntilValueChangedFires();

value2 = await directory2.get("key");  // new-value
```

I'm not sure what other alternatives exist, since we need to properly test the case where `container2.disconnect()` fails. In this case `value1` and `value2` should be equal, and we need to give enough time for the changes to sync.

Reply:
As discussed in person - This approach makes sense for this negative test, and the half second delay is ok for an e2e test.